### PR TITLE
Summarization

### PR DESCRIPTION
**Pull request by 2Rings**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by 2Rings** [comment]
Mainly about data preprocessing.

**Comment by 2Rings** [comment]
Add data2idx

File: scripts/text_summarization/data/make_datafiles.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    # tokenize_stories(cnn_stories_dir, cnn_tokenized_stories_dir)
+    # tokenize_stories(dm_stories_dir, dm_tokenized_stories_dir)
+
+    write2file(all_train_urls, os.path.join(finished_files_dir, "train.txt"))
+    wrtie2file(all_val, os.path.join(finished_files_dir, "val.txt"))
+    wrtie2file(all_test, os.path.join(finished_files_dir, "test.txt"))
+
+
+
+# NotImplemented
```
Comment (cgraywang) [inline scripts/text_summarization/data/make_datafiles.py]: delete the unused code

File: scripts/text_summarization/data/make_datafiles.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        else:
+            article_lines.append(line)
+
+    article = ' '.join(article_lines)
+
+    abstract = ' '.join("%s %s %s" % (BOS, sent, EOS) for sent in highlights)
+
+    return article, abstract
+
+def write2file(url_file, out_file):
```
Comment (cgraywang) [inline scripts/text_summarization/data/make_datafiles.py]: delete the processed datasets in the PR. eventually it will goes to S3.

File: scripts/text_summarization/encoder_decoder.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
+import mxnet as mx
+from mxnet.gluon import nn, rnn
+from gluonnlp.model.attention_cell import MLPAttentionCell
+
+class Seq2SeqEncoder(Block):
+    pass
+
+class SUMEncoder(S2Seq2SeqEncoder):
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: This class is the seq2seq+attention, right?

**Comment by 2Rings** [comment]
Already deleted used codes and add loss.py as well as decode.py.

**Comment by 2Rings** [comment]
Removed datafile and seq2seq debug

**Comment by 2Rings** [comment]
change data transformer

**Comment by 2Rings** [comment]
fixed context vector shape and beamsearch

**Comment by 2Rings** [comment]
Update seq2seq + attention

File: scripts/text_summarization/encoder_decoder.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
+import mxnet as mx
+from functools import partial
+from mxnet.base import _as_list
+from mxnet.gluon import nn, rnn
+from gluonnlp.model.attention_cell import MLPAttentionCell
+from mxnet.gluon.block import Block, HybridBlock
+
+
+__all__ = ['Seq2SeqEncoder', 'Seq2Seq2SeqDecoder', 'SUMEncoder', 'SUMDecoder', 'get_summ_encoder_decoder']
+class Seq2SeqEncoder(Block):
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: Remove Seq2SeqEncoder

File: scripts/text_summarization/encoder_decoder.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+        raise NotImplementedError
+
+class SUMEncoder(Seq2SeqEncoder):
+    def __init__(self, cell_type = 'lstm', hidden_size = 128,
+                dropout = 0.0, i2h_weight_initializer = None, h2h_weight_initializer = None,
+                i2h_bias_initializer = 'zeros', h2h_bias_initializer = 'zeros',
+                prefix = None, params = None):
+        super(SUMEncoder, self).__init__(prefix = prefix, params = params)
+        self.hidden_size = hidden_size
+        with self.name_scope():
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: Use LSTM layer instead: http://mxnet.incubator.apache.org/versions/master/api/python/gluon/rnn.html?highlight=rnn#mxnet.gluon.rnn.LSTM , and use http://mxnet.incubator.apache.org/versions/master/api/python/gluon/gluon.html?highlight=sequential#mxnet.gluon.nn.Sequential

File: scripts/text_summarization/encoder_decoder.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+            inputs: (batch_size, art_sequence_length, embedding_dim)
+            states: list of NDArray or None
+                Initial States. The list of initial states
+            return:
+            outputs: (batch_size, sequence_length, 2*num_hidden)
+            new_state: lists of new_state
+        """
+        _, length, _ = inputs.shape
+
+        outputs, new_state = self.rnn_cells[0].unroll(
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: layout = 'TNC'

Comment (cgraywang) [reply scripts/text_summarization/encoder_decoder.py]: instead of unroll, use rnn forward, and the output and output_states will be generated accordingly

File: scripts/text_summarization/encoder_decoder.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+            length = length, inputs = inputs, begin_state = states, merge_outputs = True,
+            valid_length = valid_length, layout = 'NTC')
+
+        return [outputs, new_state]
+
+class Attention(HybridBlock):
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: Ignore convolution in the encoder for attention for now

File: scripts/text_summarization/encoder_decoder.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        attn_dist = self.masked_attention(e)
+
+        context_vector = mx.ndarray.sum(mx.ndarray.reshape(attn_dist, [batch_size, -1, 1, 1]) * encoder_states, [1, 2])
+        context_vector = mx.ndarray.reshape(context_vector, [-1, attn_dist])
+
+        return context_vector, attn_dist
+
+
+
+class Seq2SeqDecoder(Block):
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: Remove Seq2SeqDecoder

File: scripts/text_summarization/encoder_decoder.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    def decode_seq(self, inputs, states, valid_length = None):
+        raise NotImplementedError
+
+    def __call__(self, step_input, states):
+        return self.forward(step_input, states)
+
+    def forward(self, step_input, states):
+        raise NotImplementedError
+
+class SUMDecoder(Seq2SeqDecoder):
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: Change the name

File: scripts/text_summarization/encoder_decoder.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+        super(SUMDecoder, self).__init__(prefix = prefix, params = params)
+        self._cell_type = rnn.LSTMCell
+        self._num_bi_layers = num_bi_layers
+        self._hidden_size = hidden_size
+        self.vocab = vocab
+        self._embedding_size = embedding_size
+
+        print("vocab_lenght: ", len(self.vocab))
+        self.attention_cell = MLPAttentionCell(units=2*self._hidden_size, normalized=False, prefix= 'attention_')
+        with self.name_scope():
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: change to LSTM layer similar to encoder

File: scripts/text_summarization/encoder_decoder.py [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                self._cell_type(hidden_size = self._hidden_size,
+                i2h_weight_initializer = i2h_weight_initializer,
+                h2h_weight_initializer = h2h_weight_initializer,
+                i2h_bias_initializer = i2h_bias_initializer,
+                h2h_bias_initializer = h2h_bias_initializer,
+                prefix = 'dec_rnn_'
+                )
+            )
+
+        with self.name_scope():
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: merge the name_scope together

File: scripts/text_summarization/encoder_decoder.py (hunk 9 above)
Comment (cgraywang) [reply scripts/text_summarization/encoder_decoder.py]: Decode function: use attention and encoder output to compute output

File: scripts/text_summarization/loss.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import numpy as np
```
Comment (cgraywang) [inline scripts/text_summarization/loss.py:1]: Remove sequence loss

File: scripts/text_summarization/run_summarization.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+else:
+    ctx = mx.gpu(args.gpu)
+
+encoder, decoder = get_summ_encoder_decoder(hidden_size=args.hidden_dim,embedding_size=args.embedding_dim, vocab=my_vocab)
+
+model = SummarizationModel(vocab = my_vocab, encoder = encoder, decoder = decoder, hidden_dim = args.hidden_dim, embed_size = args.embedding_dim, prefix = 'summary_')
+
+loss_function = SoftmaxCELoss()
+loss_function.initialize(init = mx.init.Uniform(0.02), ctx = ctx)
+loss_function.hybridize()
```
Comment (cgraywang) [inline scripts/text_summarization/run_summarization.py]: remove hybridize

File: scripts/text_summarization/run_summarization.py [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+# print "#56"
+model.initialize(init=mx.init.Uniform(0.02), ctx=ctx)
+model.hybridize()
```
Comment (cgraywang) [inline scripts/text_summarization/run_summarization.py]: remove hybridize

File: scripts/text_summarization/run_summarization.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+                log_wc = 0
+
+
+            if batch_id%500 == 0:
+                save_path = os.path.join(args.save_dir, 'valid_best.params')
+                logging.info('Save best parameters to {}'.format(save_path))
+                model.save_params(save_path)
+                # raise Exception("Save Model!")
+
+    # ## TODO: evaluation and rouge
```
Comment (cgraywang) [inline scripts/text_summarization/run_summarization.py]: add evaluation

File: scripts/text_summarization/run_summarization.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+    ctx = mx.cpu()
+    print('Use CPU')
+else:
+    ctx = mx.gpu(args.gpu)
+
+encoder, decoder = get_summ_encoder_decoder(hidden_size=args.hidden_dim,embedding_size=args.embedding_dim, vocab=my_vocab)
+
+model = SummarizationModel(vocab = my_vocab, encoder = encoder, decoder = decoder, hidden_dim = args.hidden_dim, embed_size = args.embedding_dim, prefix = 'summary_')
+
+loss_function = SoftmaxCELoss()
```
Comment (cgraywang) [inline scripts/text_summarization/run_summarization.py]: change to SoftmaxCrossEntropy

Comment (cgraywang) [reply scripts/text_summarization/run_summarization.py]: http://mxnet.incubator.apache.org/versions/master/api/python/gluon/loss.html?highlight=softmaxcrossentropy#mxnet.gluon.loss.SoftmaxCrossEntropyLoss

File: scripts/text_summarization/encoder_decoder.py [hunk 17]
```diff
@@ -0,0 +1,343 @@
+import mxnet as mx
```
Comment (cgraywang) [inline scripts/text_summarization/encoder_decoder.py]: Add test cases for encoder and decoder following https://github.com/dmlc/gluon-nlp/blob/master/tests/unittest/test_convolutional_encoder.py

**Review by cgraywang — COMMENTED** [review]
We need to improve the code

File: scripts/text_summarization/run_summarization.py (hunk 16 above)
Comment (cgraywang) [reply scripts/text_summarization/run_summarization.py]: remove bucketing, use own dataloader

**Comment by 2Rings** [comment]
update
